### PR TITLE
Support globbing filter for dotnet test

### PR DESCRIPTION
**Pull request by mariam-abdulla**
For example: 
```
dotnet test --test-modules "**/bin/**/Debug/net8.0/ConsoleApp.dll"
dotnet test --test-modules "**/bin/**/Debug/net8.0/ConsoleApp.dll" --root-directory "c:\code"
```

### CODE REVIEW

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 1]
```diff
@@ -67,28 +67,42 @@ public int Run(ParseResult parseResult)
...
-                VSTestTrace.SafeWriteTrace(() => $"MSBuild command line arguments: {string.Join(" ", msbuildCommandlineArgs)}");
-            }
+                // If the module path pattern(s) was provided, we will use that to filter the test modules
+                string testModules = parseResult.GetValue(TestCommandParser.TestModules);
 
-            ForwardingAppImplementation msBuildForwardingApp = new(GetMSBuildExePath(), msbuildCommandlineArgs);
-            int testsProjectResult = msBuildForwardingApp.Execute();
+                // If the root directory was provided, we will use that to search for the test modules
+                // Otherwise, we will use the current directory
+                string rootDirectory = parseResult.GetValue(TestCommandParser.TestModulesRootDirectory) ?? Directory.GetCurrentDirectory();
```
Comment (MarcoRossignoli) [inline src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:81]: We should check if the `TestModulesRootDirectory` exists or return a message and faile.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 2]
```diff
@@ -67,28 +67,42 @@ public int Run(ParseResult parseResult)
...
 
-            bool containsNoBuild = parseResult.UnmatchedTokens.Any(token => token == CliConstants.NoBuildOptionKey);
-            List<string> msbuildCommandlineArgs = [$"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
-                 $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
-                    "-verbosity:q"];
-
-            AddAdditionalMSBuildParameters(parseResult, msbuildCommandlineArgs);
-
-            if (VSTestTrace.TraceEnabled)
+            if (parseResult.HasOption(TestCommandParser.TestModules))
```
Comment (MarcoRossignoli) [inline src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:74]: I would refactor the body of the method inside something like `RunWithTestModulesFilter` to keep the entry point short and clear.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 3]
```diff
@@ -67,28 +67,42 @@ public int Run(ParseResult parseResult)
...
+                {
+                    VSTestTrace.SafeWriteTrace(() => $"No test modules found for the given test module pattern: {testModules} with root directory: {rootDirectory}");
+                    return 1;
+                }
 
-            if (testsProjectResult != 0)
+                foreach (string testModule in testModulePaths)
+                {
+                    _testApplications[testModule] = new TestApplication(testModule, _pipeNameDescription.Name, _args);
+                    _actionQueue.Enqueue(_testApplications[testModule]);
```
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:94]: Maybe we could update WaitAllActiojns() to return the exit code, we need to make the two branch coherent for the exit code.

**Review by MarcoRossignoli — COMMENTED** [review]
Some comments to discuss

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 4]
```diff
@@ -102,6 +116,38 @@ public int Run(ParseResult parseResult)
...
+            var testModulePatterns = testModules.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
+
+            Matcher matcher = new();
+            matcher.AddIncludePatterns(testModulePatterns);
+
+            PatternMatchingResult result = matcher.Execute(
+                new DirectoryInfoWrapper(
+                    new DirectoryInfo(rootDirectory)));
+
+            return result.Files.Select(file => $"{rootDirectory}\\{file.Path.Replace("/", "\\")}");
```
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:130]: Is it fine for Linux to normalize to backslash? It's usually easier to normalize on the most restrictive (i.e. forward slash).

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 5]
```diff
@@ -102,6 +116,38 @@ public int Run(ParseResult parseResult)
...
+        }
+
+        private int RunMSBuildTask(ParseResult parseResult)
+        {
+            bool containsNoBuild = parseResult.UnmatchedTokens.Any(token => token == CliConstants.NoBuildOptionKey);
+            List<string> msbuildCommandlineArgs = [$"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
+                 $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
+                    "-verbosity:q"];
```
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:138]:
```suggestion
            List<string> msbuildCommandlineArgs =  [
                $"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
                $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
                "-verbosity:q"
                ];
```

or

```suggestion
            List<string> msbuildCommandlineArgs =  
                [
                    $"-t:{(containsNoBuild ? string.Empty : "Build;")}_GetTestsProject",
                    $"-p:GetTestsProjectPipeName={_pipeNameDescription.Name}",
                    "-verbosity:q"
                ];
```

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 6]
```diff
@@ -67,28 +67,42 @@ public int Run(ParseResult parseResult)
...
-            int testsProjectResult = msBuildForwardingApp.Execute();
+                // If the root directory was provided, we will use that to search for the test modules
+                // Otherwise, we will use the current directory
+                string rootDirectory = parseResult.GetValue(TestCommandParser.TestModulesRootDirectory) ?? Directory.GetCurrentDirectory();
+                var testModulePaths = GetMatchedModulePaths(testModules, rootDirectory);
+
+                // If no matches were found, we simply return
+                if (!testModulePaths.Any())
+                {
+                    VSTestTrace.SafeWriteTrace(() => $"No test modules found for the given test module pattern: {testModules} with root directory: {rootDirectory}");
```
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:87]: This log is for "troubleshooting" similat to `COREHOST_TRACE=1` it's do designed for user usage.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs [hunk 7]
```diff
@@ -67,28 +67,42 @@ public int Run(ParseResult parseResult)
...
+            }
+            else
             {
-                VSTestTrace.SafeWriteTrace(() => $"MSBuild task _GetTestsProject didn't execute properly.");
-                return testsProjectResult;
+                // If no filter was provided, MSBuild will get the test project paths
+                var result = RunMSBuildTask(parseResult);
+                if (result != 0)
+                {
+                    VSTestTrace.SafeWriteTrace(() => $"MSBuild task _GetTestsProject didn't execute properly.");
```
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:103]: We can print the code, also if I don't know if is relevant it's msbuild return value.

File: src/Cli/dotnet/commands/dotnet-test/TestApplicationActionQueue.cs [hunk 8]
```diff
@@ -36,13 +40,18 @@ public void EnqueueCompleted()
...
         {
             while (await _channel.Reader.WaitToReadAsync())
             {
                 if (_channel.Reader.TryRead(out TestApplication testApp))
                 {
-                    await action(testApp);
+                    int result = await action(testApp);
+                    lock (_lock)
+                    {
+                        _hasFailed |= result != ExitCodes.Success;
```
Comment (MarcoRossignoli) [inline src/Cli/dotnet/commands/dotnet-test/TestApplicationActionQueue.cs:53]: I would remove the lock and simply check if result is not zero and set to true. Here the race is benign, if we will enter in the if we will go to set to true..so the race could read a current _hasFailed true or false...but anyway we will set to true at the end, so the state will be ok.

**Review by MarcoRossignoli — APPROVED** [review]
Some nit but LGTM

**Review by MarcoRossignoli — APPROVED** [review]

**Pull request merged**